backend/internal/oauth/oauth2/clientauth/clientauth.go

```diff
@@ -0,0 +1,117 @@
+package clientauth
+
+import (
+	"net/http"
+
+	"github.com/asgardeo/thunder/internal/application"
+	"github.com/asgardeo/thunder/internal/oauth/oauth2/constants"
+	oauth2context "github.com/asgardeo/thunder/internal/oauth/oauth2/context"
+	"github.com/asgardeo/thunder/internal/system/hash"
+	"github.com/asgardeo/thunder/internal/system/utils"
+)
+
+// Authenticate authenticates the OAuth2 client from the request.
+// It extracts credentials, validates them, and returns OAuthClientInfo on success.
+// Returns an AuthError on failure.
+func Authenticate(
+	r *http.Request,
+	appService application.ApplicationServiceInterface,
+) (*oauth2context.OAuthClientInfo, *AuthError) {
+	var clientID string
+	var clientSecret string
+	var tokenAuthMethod constants.TokenEndpointAuthMethod
+
+	// Check Authorization header (client_secret_basic)
+	if r.Header.Get("Authorization") != "" {
+		var err error
+		clientID, clientSecret, err = utils.ExtractBasicAuthCredentials(r)
+		if err != nil {
+			if err.Error() == "invalid authorization header" {
+				return nil, ErrInvalidAuthorizationHeader
+			}
+			return nil, ErrInvalidClientCredentials
+		}
+	}
+
+	// Check request body (client_secret_post)
+	clientIDFromBody := r.FormValue(constants.RequestParamClientID)
+	clientSecretFromBody := r.FormValue(constants.RequestParamClientSecret)
+
+	// Error if both header and body have credentials
+	if (clientID != "" || clientSecret != "") &&
+		(clientIDFromBody != "" || clientSecretFromBody != "") {
+		return nil, ErrBothHeaderAndBody
+	}
+
+	// Determine authentication method
+	if clientID != "" && clientSecret != "" {
+		tokenAuthMethod = constants.TokenEndpointAuthMethodClientSecretBasic
+	}
+
+	if clientIDFromBody != "" {
+		clientID = clientIDFromBody
+		if clientSecretFromBody != "" {
+			clientSecret = clientSecretFromBody
+			tokenAuthMethod = constants.TokenEndpointAuthMethodClientSecretPost
+		} else {
+			tokenAuthMethod = constants.TokenEndpointAuthMethodNone
+		}
+	}
+
+	// Validate required fields
+	if clientID == "" {
+		return nil, ErrMissingClientID
+	}
+
+	if clientSecret == "" && tokenAuthMethod != constants.TokenEndpointAuthMethodNone {
+		return nil, ErrMissingClientSecret
+	}
+
+	// Retrieve the OAuth application
+	oauthApp, err := appService.GetOAuthApplication(clientID)
+	if err != nil || oauthApp == nil {
+		return nil, ErrClientNotFound
+	}
+
+	// Validate the token endpoint authentication method
+	if !oauthApp.IsAllowedTokenEndpointAuthMethod(tokenAuthMethod) {
+		if tokenAuthMethod == constants.TokenEndpointAuthMethodNone {
+			return nil, ErrMissingClientSecret
+		}
+		return nil, ErrUnauthorizedAuthMethod
+	}
+
+	// Validate the client credentials
+	hashedClientSecret := hash.GenerateThumbprintFromString(clientSecret)
+	if tokenAuthMethod != constants.TokenEndpointAuthMethodNone {
+		if clientID != oauthApp.ClientID ||
+			hashedClientSecret != oauthApp.HashedClientSecret {
+			return nil, ErrInvalidClientSecret
+		}
+	}
```

Comment:
doesn't the oauthApp has a functio to handle this?
```
oauthApp.validateClientSecret(clientSecret)
```

This logic should be part of oauthApp not to be handled outside